### ISSUE TITLE
gif not animated

### ISSUE THREAD

**Issue by ourgit** [koral--/android-gif-drawable#260]
I am using android-gif-drawable for animate emotion. However it only shows still frame,the gif can not animate.I am sure the gif file is correct,it can animate when using Glide.

Here is the snippet:

 public static SpannableStringBuilder parseEmotion(Context context, String content, final TextView gifTextView) {
        SpannableStringBuilder ssb = new SpannableStringBuilder(content);
        Pattern pattern = Pattern.compile(FACE_PATTERN, Pattern.CASE_INSENSITIVE);
        Log.i(TAG, "before parse：" + content);
        if (null == content || content.length() < 1) return null;
        Matcher matcher = pattern.matcher(content);
        if (matcher.groupCount() > 0) {
            gifCallback = new Drawable.Callback() {
                @Override
                public void invalidateDrawable(Drawable drawable) {
                    gifTextView.invalidate();
                }

                @Override
                public void scheduleDrawable(Drawable who, Runnable what, long when) {
                    gifTextView.postDelayed(what, when);
                }

                @Override
                public void unscheduleDrawable(Drawable who, Runnable what) {
                    gifTextView.removeCallbacks(what);
                }
            };
        }

        while (matcher.find()) {
            int matchStart = matcher.start();
            int matchEnd = matcher.end();
            String group = matcher.group();
            try {
                int resourceId = context.getResources().getIdentifier(group, "drawable", "com.mypack");
                GifDrawable gifDrawable = new GifDrawable(context.getResources(), resourceId);
                gifs.add(gifDrawable);
                int width = getRecommendScaleOnBig(context);
                gifDrawable.setBounds(0, 0, width, width);
                gifDrawable.start();
                gifDrawable.setLoopCount(0);
                ssb.setSpan(new ImageSpan(gifDrawable), matchStart, matchEnd, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
                gifDrawable.setCallback(gifCallback);
            } catch (IOException e) {
                Log.e(TAG, e.getMessage());
            }
        }

        return ssb;
    }

hope guys help me out.thanks in advance

**Comment by ourgit** [comment]
I also log the method invalidateDrawable,and find it does not execute when listview nofitydataInvalidate.  I am trying to  execute drawable.invalidateSelf() to make it animate,but failed.

**Comment by koral--** [comment]
Does it animate after calling `gifTextView.setLayerType(View.LAYER_TYPE_SOFTWARE, null);`?

**Comment by ourgit** [comment]
where should I place gifTextView.setLayerType(View.LAYER_TYPE_SOFTWARE, null)? 

I placed above "return ssb",no work; 
placed      here
           gifDrawable.addAnimationListener(new AnimationListener() {
                   @Override
                 public void onAnimationCompleted(int loopNumber) {
                      Log.i(TAG,"reseting gif...");
                       gifTextView.setLayerType(View.LAYER_TYPE_SOFTWARE, null);
                      gifDrawable.invalidateSelf();
                   }
             });

some animate,some not?

where is the proper place to put the method?

thanks

**Comment by ourgit** [comment]
Hi koral,

where should I place gifTextView.setLayerType(View.LAYER_TYPE_SOFTWARE,
null)?

I placed above "return ssb",no work;
placed here
gifDrawable.addAnimationListener(new AnimationListener() {
@Override <https://github.com/Override>
public void onAnimationCompleted(int loopNumber) {
Log.i(TAG,"reseting gif...");
gifTextView.setLayerType(View.LAYER_TYPE_SOFTWARE, null);
gifDrawable.invalidateSelf();
}
});

some animate,some not. only one item animated,other item does not.

where is the proper place to put the method?

thanks

**Comment by ourgit** [comment]
I attach the gif here,maybe it could do some help for excluding error.

**Comment by koral--** [comment]
`setLayerType()` should be called once, preferably somewhere at the beginning, eg. just after retrieving `TextView` instance by `findViewById()` or similar call.
There is a similar issue, but with no solution yet: https://github.com/koral--/android-gif-drawable/issues/247
Could you tell what is the device model, which Android version it is using?
BTW there is no gif attached.

**Comment by ourgit** [comment]
I test in two device,both with the same result,one is samsung S5 with android 5.0 ,the other is samsung A5 with android 4.4.4.

git does not accept rar attachment,so I mail you with gmail attachment,maybe it can find the attachment at gmail.If need further infomation,please let me know.

**Comment by ourgit** [comment]
Hi @koral-- 
I try your advice,and it works,bingo! However I find that when I scroll the listview the memory does not release,so it keep high using memory.That is really a headache,any solution for that?thanks

**Comment by koral--** [comment]
Assuming that `parseEmotion()` is called on each item instantiation you need to unset callback (`gifDrawable.setCallback(null);`) on recycled items.

**Comment by ourgit** [comment]
yeah,parseEmotion() is called on each item instantiation. And every textview cached by convertview cache. I will try unset callback to release memory.Do you have any idea about "it works most of the Gif emotion,but some still no animate,and sometimes the not animated gifs animate when I scroll the listView"?

**Comment by koral--** [comment]
Without debugging I don't know yet what is the reason.

I think we have to first determine if callback methods (those from `Drawable.Callback`) are called for affected items when they are not animating. To check this you can eg. create (for this experiment purposes only) separate callback instances for each item. Something like that:
```java
gifDrawable.setCallback(
new Drawable.Callback() {
@Override
public void invalidateDrawable(Drawable drawable) {
gifTextView.invalidate();
//log item index to logcat, set breakpoint with log or something similar here
}
            @Override
            public void scheduleDrawable(Drawable who, Runnable what, long when) {
                gifTextView.postDelayed(what, when);
            }

            @Override
            public void unscheduleDrawable(Drawable who, Runnable what) {
                gifTextView.removeCallbacks(what);
            }
        }
);
```
After we know if callbacks are called we can decide what should be next steps.

**Comment by ourgit** [comment]
hi @koral-- 
after set and read the log, I am sure the callbacks are called. hope you can give me the further solution.
here are some key code,maybe it will do some help for solve the issue.
` gifDrawable.setCallback(new GifCallback(gifTextView));`

```
 private static class GifCallback implements Drawable.Callback {

        private final TextView tv;

        public GifCallback(TextView tv) {
            this.tv = tv;
        }

        @Override
        public void invalidateDrawable(android.graphics.drawable.Drawable who) {
            tv.invalidate();
            Log.i(TAG,"invalidateDrawable:"+tv.toString()+":drawables:"+who.toString());
        }

        @Override
        public void scheduleDrawable(android.graphics.drawable.Drawable who, Runnable what, long when) {
            tv.postDelayed(what, when);
        }

        @Override
        public void unscheduleDrawable(android.graphics.drawable.Drawable who, Runnable what) {
            tv.removeCallbacks(what);
        }
    }
```

**Comment by koral--** [comment]
How exactly are you using `parseEmotion2()` and are there any pattern which items are not animating?
There is an issue with `ListView`/`GridView` that at certain indexes are instantiated multiple times and the latest instance is not visible: https://github.com/koral--/android-gif-drawable/issues/135

**Comment by ourgit** [comment]
@koral-- 

```
SpannableStringBuilder span = EmotionParser.getInstance().parseEmotion2(instance,content, cache.tv_chatcontent_out_mul);
cache.tv_chatcontent_out_mul.setText(span);
```
I don't find any pattern which items not animateing, it seems like randomly happens,but can reproduce.

all the gifs by android-gif-drawable are visible except some not animate.

**Comment by koral--** [comment]
OK, I'll try to reproduce that.

**Comment by ourgit** [comment]
thank you. Look forward to your solution.

**Comment by koral--** [comment]
I've managed to reproduce that and will fix it ASAP.

**Comment by koral--** [comment]
@ourgit it seems that in your case the problem is that callback is garbage collected, because `Drawable` creates weak reference of it and there is no strong reference kept. You need to store reference to callback in some object which lives for the time when you need animation eg. inside item view holder.

TL;DR
Your case is not a bug in library per se however I've updated the sample project to show span inside RecyclerView: https://github.com/koral--/android-gif-drawable-sample/blob/master/sample/src/main/java/pl/droidsonroids/gif/sample/sources/GifSourcesAdapter.java#L47 and updated `MultiCallback` to support `TextView` with spans containing `GifDrawable`s.

**Comment by ourgit** [comment]
thank you. I will try your solution and see if it works.
